File: js/programmers/level1/실패율.js
function solution(N, stages) {
  const usersPerStage = new Array(N + 1).fill(0);
  let userCnt = 0;
  const failureRates = [];
  for (let stage of stages) {
    usersPerStage[stage - 1] += 1;
    userCnt += 1;
  }
  for (let i = 0; i < usersPerStage.length - 1; i++) {
    failureRates.push([usersPerStage[i] / userCnt, i]);
    userCnt -= usersPerStage[i];
  }
  failureRates.sort((a, b) => {
    // 실패율 내림차순
    if (a[0] < b[0]) return 1;
    else if (a[0] > b[0]) return -1;
    // stage 오름차순
    else if (a[1] > b[1]) return 1;
    else if (a[1] < b[1]) return -1;
  });
  return failureRates.map((v) => {
    return v[1] + 1;
  });
}

console.log(solution(5, [2, 1, 2, 6, 2, 4, 3, 3]));
console.log(solution(4, [4, 4, 4, 4, 4]));

File: 프로그래머스/lv1/42889. 실패율/실패율.py
# 실패율 = 도달했으나 클리어 x / 스테이지 도달한 플레이어 수

# 실패율이 높은 스테이지부터 내림차순으로 스테이지의 번호가 담겨있는 배열을 리턴


def solution(N, stages):
    data = [0]*(N+2)
    result = []
    l = len(stages)
    for stage in stages:
        data[stage] += 1
    k = 0
    for i in range(1, N+1):
        if l-k == 0:        
            result.append([0, i])
        else:
            result.append([data[i]/(l-k) ,i])
        k += data[i]
    result.sort(key=lambda x: -x[0])
    result = [x[1] for x in result]    
    return result